### ISSUE TITLE
'tcg fatal error' from QEMU when trying to emulate non-sense FF D8 opcode

### ISSUE THREAD

**Issue by psrok1** [unicorn-engine/unicorn#1716]
Hi!

One of our tools is using Unicorn for shellcode detection. While feeding it with arbitrary data, we noticed that QEMU used by Unicorn is crashing on `FF D8` opcode. These bytes come from the beginning of JPEG magic (`\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x01\x01\x00`)

`FF D8` is invalid instruction that is parsed as `CALL FAR m16:32` with mod=3, rm=0 (`CALL FAR rax`?) and normally results in #UD.

Is it expected to trigger `tcg_abort` on such non-sense code or is it indicator of a bug in QEMU? 

Output from debug build executed with `UNICORN_DEBUG=1` under gdb:
```
*** TCG before optimization:
 0:  ld_i32 tmp11,env,$0xfffffffffffffff0
 1:  movi_i32 tmp12,$0x0
 2:  brcond_i32 tmp11,tmp12,lt,$L0

 insn_idx=0 ---- 0000000000100000 0000000000000000
 1:  movi_i64 tmp3,$0x100000
 2:  st_i64 tmp3,env,$0x80
 3:  movi_i32 tmp11,$0x2
 4:  movi_i64 tmp13,$0x555555fcfe50
 5:  movi_i64 tmp14,$0x100000
 6:  movi_i32 tmp12,$0x2
 7:  call uc_tracecode,$0x0,$0,tmp11,tmp12,tmp13,tmp14
 8:  ld_i32 tmp11,env,$0xfffffffffffffff0
 9:  movi_i32 tmp12,$0x0
 10:  brcond_i32 tmp11,tmp12,lt,$L0
 11:  mov_i64 tmp0,rax
 12:  qemu_ld_i64 tmp1,tmp2,leul,2
 13:  ld_i32 tmp11,env,$0xfffffffffffffff0
 14:  movi_i32 tmp12,$0x0
 15:  brcond_i32 tmp11,tmp12,lt,$L0
 16:  movi_i64 tmp13,$0x4
 17:  add_i64 tmp2,tmp2,tmp13
 18:  qemu_ld_i64 tmp0,tmp2,leuw,2
 19:  ld_i32 tmp11,env,$0xfffffffffffffff0
 20:  movi_i32 tmp12,$0x0
 21:  brcond_i32 tmp11,tmp12,lt,$L0
 22:  mov_i32 tmp5,tmp0
 23:  movi_i64 tmp13,$0x100002
 24:  movi_i32 tmp11,$0x1
 25:  call lcall_protected,$0x0,$0,env,tmp5,tmp1,tmp11,tmp13
 26:  ld_i64 tmp7,env,$0x80
 27:  call lookup_tb_ptr,$0x6,$1,tmp14,env
 28:  goto_ptr tmp14
 29:  set_label $L0
 30:  exit_tb $0x7fffb2d42043

*** TCG before codegen:
 0:  ld_i32 tmp11,env,$0xfffffffffffffff0
 1:  movi_i32 tmp12,$0x0
 2:  brcond_i32 tmp11,tmp12,lt,$L0  dead: 0 1

 insn_idx=0 ---- 0000000000100000 0000000000000000
 1:  movi_i64 tmp3,$0x100000
 2:  st_i64 tmp3,env,$0x80  dead: 0 1
 3:  movi_i32 tmp11,$0x2
 4:  movi_i64 tmp13,$0x555555fcfe50
 5:  movi_i64 tmp14,$0x100000
 6:  movi_i32 tmp12,$0x2
 7:  call uc_tracecode,$0x0,$0,tmp11,tmp12,tmp13,tmp14  dead: 0 1 2 3
 8:  ld_i32 tmp11,env,$0xfffffffffffffff0
 9:  movi_i32 tmp12,$0x0
 10:  brcond_i32 tmp11,tmp12,lt,$L0  dead: 0 1
 11:  qemu_ld_i64 tmp1,tmp2,leul,2
 12:  ld_i32 tmp11,env,$0xfffffffffffffff0
 13:  movi_i32 tmp12,$0x0
 14:  brcond_i32 tmp11,tmp12,lt,$L0  dead: 0 1
 15:  movi_i64 tmp13,$0x4
 16:  add_i64 tmp2,tmp2,tmp13  dead: 1 2
 17:  qemu_ld_i64 tmp0,tmp2,leuw,2  dead: 1
 18:  ld_i32 tmp11,env,$0xfffffffffffffff0
 19:  movi_i32 tmp12,$0x0
 20:  brcond_i32 tmp11,tmp12,lt,$L0  dead: 0 1
 21:  mov_i32 tmp5,tmp0  dead: 1
 22:  movi_i64 tmp13,$0x100002
 23:  movi_i32 tmp11,$0x1
 24:  call lcall_protected,$0x0,$0,env,tmp5,tmp1,tmp11,tmp13  dead: 0 1 2 3 4
 25:  call lookup_tb_ptr,$0x6,$1,tmp14,env  dead: 1
 26:  goto_ptr tmp14  dead: 0
 27:  set_label $L0
 28:  exit_tb $0x7fffb2d42043
/home/psrok1/sflock/unicorn/qemu/tcg/tcg.c:3073: tcg fatal error

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350254592) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350254592) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140737350254592) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140737350254592, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff7c8f476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff7c757f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007ffff40aec66 in temp_load (s=0x55555606bda0, ts=0x55555606c978, desired_regs=65343, allocated_regs=48, preferred_regs=0) at /home/psrok1/sflock/unicorn/qemu/tcg/tcg.c:3073
#6  0x00007ffff40afa85 in tcg_reg_alloc_op (s=0x55555606bda0, op=0x555555ff9780) at /home/psrok1/sflock/unicorn/qemu/tcg/tcg.c:3455
#7  0x00007ffff40b0ae2 in tcg_gen_code_x86_64 (s=0x55555606bda0, tb=0x7fffb2d42040) at /home/psrok1/sflock/unicorn/qemu/tcg/tcg.c:3834
#8  0x00007ffff40dabb3 in tb_gen_code_x86_64 (cpu=0x5555560ae7a0, pc=1048576, cs_base=0, flags=4243632, cflags=-16777216) at /home/psrok1/sflock/unicorn/qemu/accel/tcg/translate-all.c:1648
#9  0x00007ffff40c3b12 in tb_find (cpu=0x5555560ae7a0, last_tb=0x0, tb_exit=0, cf_mask=0) at /home/psrok1/sflock/unicorn/qemu/accel/tcg/cpu-exec.c:256
#10 0x00007ffff40c4406 in cpu_exec_x86_64 (uc=0x555555fcfe50, cpu=0x5555560ae7a0) at /home/psrok1/sflock/unicorn/qemu/accel/tcg/cpu-exec.c:597
#11 0x00007ffff4086873 in tcg_cpu_exec (uc=0x555555fcfe50) at /home/psrok1/sflock/unicorn/qemu/softmmu/cpus.c:96
#12 0x00007ffff4086b2c in resume_all_vcpus_x86_64 (uc=0x555555fcfe50) at /home/psrok1/sflock/unicorn/qemu/softmmu/cpus.c:215
#13 0x00007ffff4086bc7 in vm_start_x86_64 (uc=0x555555fcfe50) at /home/psrok1/sflock/unicorn/qemu/softmmu/cpus.c:234
#14 0x00007ffff4072f93 in uc_emu_start (uc=0x555555fcfe50, begin=1048576, until=1048578, timeout=0, count=1048576) at /home/psrok1/sflock/unicorn/uc.c:870
```

We can see that our instruction was interpreted as far call: `24:  call lcall_protected,$0x0,$0,env,tmp5,tmp1,tmp11,tmp13  dead: 0 1 2 3 4`

**Comment by wtdcode** [comment]
Hello, sorry for the late reply. It's a Unicorn's bug I think because anyway we shouldn't abort in such cases...

**Comment by kem0x** [comment]
@wtdcode any plan to get this fixed soon? i'm having the same issue here i believe...

**Comment by CR3Swapper** [comment]
same issue here, unicorn engine exploding when throwing random bytes/instructions at it. I assume its a qemu/underlying tcg issue. is this something the unicorn engine team will be interested in fixing possibly?

**Comment by wtdcode** [comment]
I believe it should be related to #1869 where I have an incomplete fix. Still working on these two issues and sorry for the delay.

**Comment by kem0x** [comment]
> I believe it should be related to #1869 where I have an incomplete fix. Still working on these two issues and sorry for the delay.

Thanks for the quick reply, hoping to hear from you more about this soon. :pray:

**Comment by wtdcode** [comment]
Oh wait this should be addressed by #1842 and I gave it a try with emulating `(\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x01\x01\x001 it should work now.

@kem0x @CR3Swapper Please move to #1869 for possible further discussions.
